Add reconfigure to jellyfin

Erotemic opened the pull request:
## Proposed change

Currently the Jellyfin integration cannot be reconfigured. You cannot change your username / password or IP address for the server. Following the [config-entry-reconfigure-step](https://developers.home-assistant.io/blog/2024/03/21/config-entry-reconfigure-step/)  I've attempted to add this feature.


## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

I am new to developing on this repo. I am using this as an opportunity to learn a bit about the core architecture. Please bear with me as I familiarize myself with standards and practices here.

I used ChatGPT to write the new code. I believe I understand what is happening and it does seem to work, but there's always a chance AI code is incorrect, and I'm not experienced enough with this repo to catch everything. 

I'm also having issues with the tests. I've been trying to follow other reconfigure tests I've seen for other components, but there is a lot of mocking infrastructure I'm not familiar with. Any help with writing these tests correctly would be appreciated, but I'll try to work through it as well. 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.


To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Erotemic committed:
Add reconfigure to jellyfin

frenck commented:
Oh hi there @Erotemic 👋 

Small reminder: this PR is currently in draft, and the CI is failing. Could you take a look and mark the PR ready for review afterward?

Thanks 👍 

../Frenck

Erotemic commented:
I still don't have a great grasp on how to get the mocks working correctly for tests. Any guidance here would be appreciated.

Hoping to figure this out in the next week or so. This is step 1 on tackling a larger issue I'm experiencing with the Jellyfin integration: namely it doesn't play any of my media. I can browse my library, but when I click play, it just hangs forever.

Erotemic commented:
So I'm starting to understand the mocks a bit better. I have the tests working IF they are run individually, but if they are all run together the invalid credentials test fails, and I cannot figure out why. First I thought that the `mock_client` object was reused between tests, so any of its return value or side effect patches from previous tests might still be active, but that doesn't seem to be the case, it has a unique id in each test, and the values on `mock_client.auth.connect_to_address.return_value` and `mock_client.auth.login.return_value` are what I expect them to be.

I feel like I must be close to passing tests, and I can probably hack it, but again as I don't often do network programming with mock testing, and because there was a sizable reliance on AI to generate this patch,  a careful review is needed.

Erotemic commented:
Looking closer, I was getting an identical version of the mock, but that wasn't the core issue. The main problem was that I was patching `homeassistant.components.jellyfin.client_wrapper.create_client` and not `homeassistant.components.jellyfin.config_flow.create_client`, which means create_client was modified in its original namespace, but by that time the original version of `create_client` was already in the config_flow namespace and patch can't know (without a big runtime cost) it needs to modify it there too.

In any case, the tests are all passing now, and I think they are testing useful things. I do expect that I'm not using best practices in some areas, so please point out where I've made errors. I'm very interested in becoming an expert on this codebase.

joostlek commented on homeassistant/components/jellyfin/config_flow.py:
```diff
@@ -135,6 +135,45 @@ async def async_step_reauth_confirm(
+    ) -> ConfigFlowResult:
+        """Allow users to update their Jellyfin configuration."""
+        errors: dict[str, str] = {}
+        entry = self.hass.config_entries.async_get_entry(self.context["entry_id"])
+
+        if not entry:
+            return self.async_abort(reason="entry_not_found")
```
You can use `self._get_reconfigure_entry()` for this

Erotemic replied:
I'm assuming you are referring to line 143 and 145-146. 

Looking at the code, I see that `_get_reconfigure_entry` calls `self.hass.config_entries.async_get_known_entry(self._reconfigure_entry_id)`, which is different than `async_get_entry` only because it will raise `raise UnknownEntry(entry_id)` is entry is None. So that will prevent the 145/146 path from ever being taken if entry is fasley, because the exception will be thrown instead.

Does some upstream mechanism handle the `UnknownEntry` exception in the same way as an abort return value?

joostlek replied:
But the thing I'm missing is, why would the entry be None

emontnemery replied:
@Erotemic the idea is that it's not possible to start a reconfigure flow without a config entry, and if the config entry is removed while the reconfigure flow is in progress, the reconfigure flow will automatically be closed. Hence, the the helper `_get_reconfigure_entry` will always return a `ConfigEntry`, avoiding useless error handling in the reconfigure flow.

joostlek commented on homeassistant/components/jellyfin/config_flow.py:
```diff
@@ -135,6 +135,45 @@ async def async_step_reauth_confirm(
+        if user_input is not None:
+            new_input = entry.data | user_input
+
+            if self.client_device_id is None:
+                self.client_device_id = new_input.get(
+                    CONF_CLIENT_DEVICE_ID, _generate_client_device_id()
+                )
```
where would this be set?

Erotemic replied:
I'm not sure what "this" means: do you mean `self.client_device_id`?

I think the idea is that we need a client ID to create an instance of the `JellyfinClient`. I'm not sure if its possible `self.client_device_id` to be None in this method. Because I'm still learning this codebase, I'm mostly just following patterns I see. I'm also not sure if `new_input` can ever have an entry for `CONF_CLIENT_DEVICE_ID`, but there is a  line in `JellyfinConfigFlow.async_step_user` where it looks like it creates an entry with an ID populated for that field:

```python
                return self.async_create_entry(
                    title=entry_title,
                    data={CONF_CLIENT_DEVICE_ID: self.client_device_id, **user_input},
                )
```

And `new_input` does look like it can be pre-populated from data that has been preconfigured, so maybe the different code paths here do make sense?

I can try removing it and see what happens. When I run the reconfigure step manually I don't see any user input for this, so it's not user facing. This is the dialog I see when I hit "reconfigure":

![image](https://github.com/user-attachments/assets/f0cb90b9-e5ea-46cf-a260-8874b7a0cb99)

Erotemic replied:
So removing this block of code doesn't seem to cause the test to fail, and the dialog still seems to work when I use it. Maybe it makes sense to ignore looking up an existing id in the `new_input` dict? I've made that change, which should at least prevent the corner case where it `self.client_device_id` is None, but I'm also not convinced it ever will be None there, nor do I know what I could do in the UI to test it.

joostlek commented on tests/components/jellyfin/test_config_flow.py:
```diff
@@ -466,3 +466,119 @@ async def test_setting_codec(
+    """Test successful reconfigure flow."""
+    mock_config_entry.add_to_hass(hass)
+
+    result0 = await hass.config_entries.options.async_init(mock_config_entry.entry_id)
+    assert result0["type"] == FlowResultType.FORM
+    assert result0["step_id"] == "init"
```
we don't need to start an options flow here

Erotemic replied:
I see, so `options.async_init` is what is starting an options flow dialog to "initialize" the options. Removed.

joostlek commented on tests/components/jellyfin/test_config_flow.py:
```diff
@@ -466,3 +466,119 @@ async def test_setting_codec(
+    assert result1["type"] is FlowResultType.FORM
+    assert result1["step_id"] == "reconfigure"
+
+    # NOTE: There is likely a better way to test a successful reconfigure where
+    # the validate input is run and its components are mocked.
+    # I don't understand why test_reauth and test_form work when they call
+    # validate input, but here if we remove the patch, it causes an error in
+    # the connection_manager where it fails to connect
```
Could it be because the rest have `mock_jellyfin`?

Erotemic replied:
Aahh. I see it now.

Man, I am not a fan of pytest fixtures :soap: :blue_square:. When I see an unused argument to a function I assume it doesn't mean anything, so I didn't copy it over.  In any case, I'm glad I understand what's going on now.

joostlek commented on tests/components/jellyfin/test_config_flow.py:
```diff
@@ -466,3 +466,119 @@ async def test_setting_codec(
+                CONF_PASSWORD: "new_password",
+            },
+        )
+    await hass.async_block_till_done()
```
```suggestion
```

Erotemic replied:
Removing these doesn't seem to break anything, but every other test in this file has them. Is there a reason this call is not needed here?

joostlek replied:
They are generally used for when code fires events, and this code doesn't. So at this point I'm just slowly removing them or stopping new ones to happen

joostlek commented on tests/components/jellyfin/test_config_flow.py:
```diff
@@ -466,3 +466,119 @@ async def test_setting_codec(
+
+        assert result2["type"] is FlowResultType.FORM
+        assert result2["errors"] == {"base": "cannot_connect"}
+
+
+async def test_reconfigure_flow_exception_invalid_credentials(
+    hass: HomeAssistant, mock_config_entry: MockConfigEntry, mock_client: MagicMock
+) -> None:
+    """Test reconfigure flow authentication failure ."""
+    mock_config_entry.add_to_hass(hass)
+
+    mock_client.auth.connect_to_address.return_value = await async_load_json_fixture(
+        hass, "auth-connect-address.json"
+    )
+    mock_client.auth.login.return_value = await async_load_json_fixture(
+        hass, "auth-login-failure.json"
+    )
+    # Force the new client to use our mock. I do not know if this is
+    # the correct way to to run this test, or if there is a better way.
+    with patch(
+        "homeassistant.components.jellyfin.config_flow.create_client",
+        return_value=mock_client,
+    ):
+        result = await mock_config_entry.start_reconfigure_flow(hass)
+
+        assert result["type"] is FlowResultType.FORM
+        assert result["step_id"] == "reconfigure"
+
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                CONF_URL: "http://new_url:8096",
+                CONF_USERNAME: "new_user",
+                CONF_PASSWORD: "invalid_new_password",
+            },
+        )
+        await hass.async_block_till_done()
+
+        assert result2["type"] is FlowResultType.FORM
+        assert result2["errors"] == {"base": "invalid_auth"}
```
You can use `pytest.mark.parametrize` to add parameters and to deduplicate these tests. We also want a test to cover the bare `Exception`.

Can we also finish these tests so we also test that the flow is able to recover?

Erotemic replied:
I made the switch to `pytest.mark.parametrize`. I generally avoid it due to aforementioned issues with pytest's implicitness, but that's not really relevant given that's what this project uses. I suppose I just can't help myself from making a comment about it.

I'm not exactly sure what you mean by "finish these tests so we also test that the flow is able to recover". The other reconfigure tests that I looked at don't seem to do aything after testing the result has the expected error message. Is testing that the `mock_config_entry.data` hasn't changed what you meant by "recover"? I put that in, but if that's not what you mean, I could use clarification.

I also added a test that covers the bare Exception. I wasn't sure how to do it with `async_load_json_fixture`, and I saw other examples where they just forced a specific exception as a side effect, so I went that route, and also added the other two exceptions explicitly. I think it makes sense to have this (which I've called `test_reconfigure_flow_exceptions`) as a separate test from the json fixture route (which I renamed to `test_reconfigure_flow_failure`) as that seems to be testing the that a specific message causes a specific exception, and this is just testing "what if an exception occurred".

Let me know if my thought process makes sense here.

joostlek replied:
With recover I mean, what if I fill in an invalid token. And then I get an error back, and if I then enter the correct values, there should not be a side effect that now effectively stops me from submitting my form. Maybe the step uses stale data or a boolean is flipped. Not that the code does that now, but by testing this we avoid that from happening

Erotemic committed:
refactor: use _get_reconfigure_entry helper

Erotemic committed:
refactor: replace manual patch with jellyfin fixture

Erotemic committed:
refactor: remove hass.async_block_till_done lines

Erotemic committed:
refactor: use pytest fixtures

Erotemic committed:
add: test bare reconfigure exceptions

Erotemic committed:
Removed unused params

Erotemic committed:
Simplify self.client_device_id is None case

Erotemic commented:
I've addressed or responded to all of the comments. I've also rebased on the latest "dev" branch. PTAL and let me know if there is more polish needed.

joostlek commented on homeassistant/components/jellyfin/config_flow.py:
```diff
@@ -135,6 +140,43 @@ async def async_step_reauth_confirm(
+    ) -> ConfigFlowResult:
+        """Allow users to update their Jellyfin configuration."""
+        errors: dict[str, str] = {}
+        entry: ConfigEntry = self._get_reconfigure_entry()
+
+        if not entry:
+            return self.async_abort(reason="entry_not_found")
```
The entry is there for sure :)

joostlek commented on tests/components/jellyfin/test_config_flow.py:
```diff
@@ -466,3 +467,131 @@ async def test_setting_codec(
+    """Test successful reconfigure flow."""
+    mock_config_entry.add_to_hass(hass)
+
+    result1 = await mock_config_entry.start_reconfigure_flow(hass)
+
+    assert result1["type"] is FlowResultType.FORM
+    assert result1["step_id"] == "reconfigure"
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result1["flow_id"],
+        {
+            CONF_URL: "http://new_url:8096",
+            CONF_USERNAME: "new_user",
+            CONF_PASSWORD: "new_password",
+        },
+    )
+    assert result2["type"] is FlowResultType.ABORT
+    assert result2["reason"] == "reconfigure_successful"
```
my general recommendation is to reuse `result` instead of appendign a number

Erotemic replied:
I generally prefer to not reassign variables when possible. I like maximizing the idempotency of code. But I agree it would be better without arbitrary numbers. What if I use `start_result` and `configure_result`? Or do you just want me to use `result` and reassign it?

joostlek commented on homeassistant/components/jellyfin/config_flow.py:
```diff
@@ -135,6 +140,43 @@ async def async_step_reauth_confirm(
+        if user_input is not None:
+            new_input = entry.data | user_input
+
+            if self.client_device_id is None:
+                self.client_device_id = _generate_client_device_id()
```
`self.client_device_id` is never set anywhere during reconfigure

Erotemic replied:
I don't understand this, there are 3 methods `async_step_reconfigure`, `async_step_reauth_confirm`, and `async_step_user` that could be called, and each of them assigns `client_device_id`, and by following other code in the repo I presume that we don't want to regenerate the device-id once we have it.

Is it the case that just this one will always be guaranteed to have `self.client_device_id` set to None? This is a limitation of my current understanding of the HA architecture. Could you help me understand?

Erotemic replied:
After consulting with ChatGPT, is it supposed to be the case that the `client_device_id` is stored in the new entry data?

Is what your asking to effectively set the client id on the entry?

```python
new_input[CONF_CLIENT_DEVICE_ID] = self.client_device_id
```

If I understand this correctly, the client-id is used to identify the connection that HA is making to jellyfin. So every time we do this reconfigure jellyfin will see a new client, but we are at least persisting that same client information within the existing ConfigFlow wizard. But if we start a new one, it will change.

Should we be grabbing an existing client-id (to prevent unnecessary re-authentication issues?) from the entry.data if it exists on a reconfigure? I.e. Instead of:

```
            if self.client_device_id is None:
                self.client_device_id = _generate_client_device_id()
```

Should I be using:

```python
            if self.client_device_id is None:
                self.client_device_id = entry.data.get(CONF_CLIENT_DEVICE_ID) or _generate_client_device_id()
```
?

joostlek commented on tests/components/jellyfin/test_config_flow.py:
```diff
@@ -466,3 +467,131 @@ async def test_setting_codec(
+    )
+    assert result2["type"] is FlowResultType.ABORT
+    assert result2["reason"] == "reconfigure_successful"
+    assert mock_config_entry.data[CONF_USERNAME] == "new_user"
+    assert mock_config_entry.data[CONF_PASSWORD] == "new_password"
+    assert mock_config_entry.data[CONF_URL] == "http://new_url:8096"
```
Let's assert the unique id as well

joostlek commented on tests/components/jellyfin/test_config_flow.py:
We should also add a test where we test that we connect to the same user/server, so can we get the user_id again and test if it matches? You can use `self._abort_if_unique_id_mismatch`

Erotemic replied:
Will try to take a look next week.

Erotemic replied:
@joostlek I've taken a pass at these requests. I'm still not confident in my understanding of the system, but I'm making progress. All of the easy requests have been done.

Beyond those, the main updates are:

* On `async_step_reconfigure` I check if the existing entry.data has a `CONF_CLIENT_DEVICE_ID` and use that instead of generating a new one.

* I update the `new_input` with the new or existing `CONF_CLIENT_DEVICE_ID` value.

* I get the `user_id` from the `valid_input` call and if it succeedes, I call `async_set_unique_id` and `_abort_if_unique_id_configured` similar to `async_step_user`. 

* I added `mock_client_device_id` to `test_reconfigure_successful` so I can test the value of `CONF_CLIENT_DEVICE_ID`

* Also updated `test_reconfigure_successful` to check that `mock_config_entry.unique_id` does not change.

* I added a new test `test_reconfigure_keep_client_id` that checks that `CONF_CLIENT_DEVICE_ID` does not change if you reconfigure multiple times. (I think this is what you were asking for, but I'm not 100% sure, please advise if not).

* I had to add `unique_id_mismatch` to strings.json to get tests to pass after I rebased on the latest dev. Not sure if this was necessary or if I did something else wrong.  

* I renamed a few other `result2` objects in my new code, although it seems like existing tests use that naming scheme, which is probably why I used it originally. I didn't touch any other tests to keep the diff minimal.

Erotemic committed:
Review: we know the entry is there

Erotemic committed:
Rename result1 and result2 to be semantically descriptive

Erotemic committed:
Test understanding of clientid

Erotemic committed:
Fix strings after rebase, improve result names, add test to ensure client id persists

Erotemic committed:
remove comment

Erotemic committed:
Merge branch 'dev' into reconfigure-jellyfin

Erotemic committed:
Update homeassistant/components/jellyfin/config_flow.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

Erotemic committed:
Merge branch 'dev' into reconfigure-jellyfin

emontnemery commented on homeassistant/components/jellyfin/config_flow.py:
```diff
@@ -135,6 +140,52 @@ async def async_step_reauth_confirm(
+                errors["base"] = "unknown"
+                _LOGGER.exception("Unexpected exception")
+            else:
+                await self.async_set_unique_id(user_id)
+                self._abort_if_unique_id_mismatch()
```
So, the point is it's possible to change the username in jellyfin, and when that happens the authentication pair user+password changes. We don't want to allow changing to a different user id though, that should instead be a new entry? If so, please add a comment explaining that in the code.

emontnemery commented on homeassistant/components/jellyfin/config_flow.py:
```diff
@@ -135,6 +140,52 @@ async def async_step_reauth_confirm(
+
+        if user_input is not None:
+            if self.client_device_id is None:
+                # Keep the same client id if we have one.
+                self.client_device_id = (
+                    entry.data.get(CONF_CLIENT_DEVICE_ID)
+                    or _generate_client_device_id()
+                )
```
Jellyfin's client list got bloated when we *never* remembered the client ID. I think it makes sense to create a new client ID when reauthenticating just to keep the code simple, and make the reauthenticate process "complete".

joostlek replied:
I do have to admit that I am not a jellyfin user, so I also don't exactly know the struggles the integration had in the past

Erotemic committed:
Update homeassistant/components/jellyfin/config_flow.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

Erotemic committed:
Merge branch 'dev' into reconfigure-jellyfin

Erotemic committed:
Undo change to keep the client id, use a new one each reauth

Erotemic committed:
Merge branch 'dev' into reconfigure-jellyfin

Erotemic committed:
Remove unused string

Erotemic committed:
Merge branch 'dev' into reconfigure-jellyfin
